lib/Reader/readerir.cpp

```diff
@@ -1634,6 +1675,84 @@ Type *GenIR::getClassType(CORINFO_CLASS_HANDLE ClassHandle, bool IsRefClass,
   return ResultTy;
 }
 
+void GenIR::addFieldsRecursively(
+    std::vector<std::pair<uint32_t, llvm::Type *>> &Fields, uint32_t Offset,
+    llvm::Type *Ty) {
+  StructType *StructTy = dyn_cast<StructType>(Ty);
+  if (StructTy != nullptr) {
+    const DataLayout *DataLayout = JitContext->EE->getDataLayout();
+    for (Type *SubTy : StructTy->subtypes()) {
+      addFieldsRecursively(Fields, Offset, SubTy);
+      Offset += DataLayout->getTypeSizeInBits(SubTy) / 8;
+    }
+  } else {
+    Fields.push_back(std::make_pair(Offset, Ty));
+  }
+}
+
+void GenIR::createOverlapFields(
+    std::vector<std::pair<uint32_t, llvm::Type *>> &OverlapFields,
+    std::vector<llvm::Type *> &Fields) {
+
+  // Prepare to create and measure types.
+  LLVMContext &LLVMContext = *JitContext->LLVMContext;
+  const DataLayout *DataLayout = JitContext->EE->getDataLayout();
+
+  // Order the OverlapFields by offset.
+  std::sort(OverlapFields.begin(), OverlapFields.end());
+
+  // Walk the fields, accumulating the unique starting offsets of the gc
+  // references in increasing offset order.
+  std::vector<uint32_t> GcOffsets;
+  uint32_t OverlapEndOffset = 0;
+  for (const auto &OverlapField : OverlapFields) {
+    uint32_t Offset = OverlapField.first;
+    uint32_t Size = DataLayout->getTypeSizeInBits(OverlapField.second) / 8;
+    OverlapEndOffset = std::max(OverlapEndOffset, Offset + Size);
+    if (isManagedPointerType(OverlapField.second)) {
+      assert(((Offset % getPointerByteSize()) == 0) &&
+             "expect aligned gc pointers");
+      if (GcOffsets.empty()) {
+        GcOffsets.push_back(Offset);
+      } else {
+        uint32_t LastOffset = GcOffsets.back();
+        assert((Offset >= LastOffset) && "expect offsets to be sorted");
+        if (Offset > LastOffset) {
+          GcOffsets.push_back(Offset);
+        }
+      }
+    }
+  }
+
+  // Walk the GC reference offsets, creating representative fields.
+  uint32_t FirstOffset = OverlapFields.begin()->first;
+  uint32_t CurrentOffset = FirstOffset;
+  for (const auto &GcOffset : GcOffsets) {
+    assert((GcOffset >= CurrentOffset) && "expect offsets to be sorted");
+    uint32_t NonGcPreambleSize = GcOffset - CurrentOffset;
+    if (NonGcPreambleSize > 0) {
+      Type *NonGcTy =
+          ArrayType::get(Type::getInt8Ty(LLVMContext), NonGcPreambleSize);
+      Fields.push_back(NonGcTy);
+    }
+    Fields.push_back(getBuiltInObjectType());
+    CurrentOffset += getPointerByteSize();
+  }
+
+  // Create a trailing non-gc field if needed.
+  uint32_t OverlapSize = OverlapEndOffset - FirstOffset;
+  uint32_t CurrentOverlapSize = CurrentOffset - FirstOffset;
+  assert((CurrentOverlapSize <= OverlapSize) && "overlap size overflow");
+  if (CurrentOverlapSize < OverlapSize) {
+    uint32_t RemainingSize = OverlapSize - CurrentOverlapSize;
```

Comment:
Seems like the five lines above could be simplified to:
```c++
  if (CurrentOffset < OverlapEndOffset) {
    uint32_t RemainingSize = OverlapEndOffset - CurrentOffset;
```
no?

(Fine as-is if you find it more readable your way)